src/sage/env.py

```diff
@@ -212,6 +212,7 @@ def var(key: str, *fallbacks: Optional[str], force: bool = False) -> Optional[st
 MAXIMA_FAS = var("MAXIMA_FAS")
 KENZO_FAS = var("KENZO_FAS")
 SAGE_NAUTY_BINS_PREFIX = var("SAGE_NAUTY_BINS_PREFIX", "")
+SAGE_ECMBIN = var("SAGE_ECMBIN")
```

Comment:
```suggestion
ECM = var("ECM")
```
... just like `MAXIMA`. Only nauty has this weird naming

Reply:
Why do we need a variable for everything under the sun? There's a well established way to configure binary locations, it's called `PATH`. If you must: wouldn't it be better to use `SAGE_*` to keep it in sage "namespace"?

As for the two standard names for the binary, that's unfortunate, is there a valid reason why Fedora does that? Is it possible/desirable to have the executable feature try both?

For non-standard setups one can always place a symlink in a custom directory and add it to PATH.

Reply:
A symlink worked and one could asked fedora packagers to place this symlink, but apparently it is not the only distribution. Concerning the error in `src/sage/libs/gap/element.pyx`, it happens in line 2489:
```
for i in range(100):
        rnd = [ randint(-10,10) for i in range(randint(0,7)) ]
        # compute the sum in GAP
        _ = libgap.Sum(rnd)
        try:
            libgap.Sum(*rnd)
            print('This should have triggered a ValueError')
            print('because Sum needs a list as argument')
        except ValueError:
            pass
```
getting `Killed due to segmentation fault`. Running this code in the terminal works, the error appears only in tests and with python 3.12.

Reply:
> A symlink worked and one could asked fedora packagers to place this symlink, but apparently it is not the only distribution.

You mean: ask the fedora packager *not to rename* the executable (https://src.fedoraproject.org/rpms/gmp-ecm/blob/rawhide/f/gmp-ecm.spec#_99). We don't need to do anything special in void linux for `ecm` to be named `ecm`. Anyway, that line there is 15 years old, and possibly other distros also use the same name for the binary, so I think it's ok to try both names (I'm only objecting to do it at configure time: just do it at runtime).

> Concerning the error in `src/sage/libs/gap/element.pyx`, it happens in line 2489:
> 
> ```
> for i in range(100):
>         rnd = [ randint(-10,10) for i in range(randint(0,7)) ]
>         # compute the sum in GAP
>         _ = libgap.Sum(rnd)
>         try:
>             libgap.Sum(*rnd)
>             print('This should have triggered a ValueError')
>             print('because Sum needs a list as argument')
>         except ValueError:
>             pass
> ```
> 
> getting `Killed due to segmentation fault`. Running this code in the terminal works, the error appears only in tests and with python 3.12.

I never saw that, I also can't see what this has to do with python 3.12. Unless something there was compiled for python 3.11 and now run on python 3.12 (since layout of integers changed). But something like that I would expect to give many more issues...

Is it something that happens every time, or at random? Did you manage to get a clean backtrace of the segfault? Are you sure the sage+gap used for tests is exactly the same you are using for trying this in the terminal?

Reply:
here is how the package got into Fedora in the first place.

https://bugzilla.redhat.com/show_bug.cgi?id=473330#c2

there is a [comment from the reviewer](
https://bugzilla.redhat.com/show_bug.cgi?id=473330#c8)
requesting the binary to be called gmp-ecm.

Reply:
> You mean: ask the fedora packager _not to rename_ the executable (https://src.fedoraproject.org/rpms/gmp-ecm/blob/rawhide/f/gmp-ecm.spec#_99). We don't need to do anything special in void linux for `ecm` to be named `ecm`. Anyway, that line there is 15 years old, and possibly other distros also use the same name for the binary, so I think it's ok to try both names (I'm only objecting to do it at configure time: just do it at runtime).

The executable name never changes so it's wasteful to check it at runtime. When sagelib gets a build system, we'll be able to get rid of the sage-conf variable, too -- ideally the name would be hard-coded at install-time.

Reply:
> Why do we need a variable for everything under the sun? 

before this PR, `ecm` is the hardcoded binary's name.

I replace it with a variable which is set up by running `./configure`.



> There's a well established way to configure binary locations, it's called `PATH`. If you must: wouldn't it be better to use `SAGE_*` to keep it in sage "namespace"?

for some reason @mkoeppe asked for this name.


> 
> As for the two standard names for the binary, that's unfortunate, is there a valid reason why Fedora does that?

Something something "but we are not Debian", as
mentioned [here] (https://bugzilla.redhat.com/show_bug.cgi?id=473330#c8)

>  Is it possible/desirable to have the executable feature try both?


You can of course do anything from Python, but this is error-prone, a security hazard, and may be slow. Besides, why disregard the results of `./configure` ?


Can Void just hardcode this variable as `"ecm"` ?


> 
> For non-standard setups one can always place a symlink in a custom directory and add it to PATH.

Reply:
> here is how the package got into Fedora in the first place.
> 
> https://bugzilla.redhat.com/show_bug.cgi?id=473330#c2
> 
> there is a [comment from the reviewer](https://bugzilla.redhat.com/show_bug.cgi?id=473330#c8) requesting the binary to be called gmp-ecm.

Fuck the NIH drug. Petty reason: "debian uses `ecm` so we should use `gmp-ecm`".

Reply:
> The executable name never changes so it's wasteful to check it at runtime. When sagelib gets a build system, we'll be able to get rid of the sage-conf variable, too -- ideally the name would be hard-coded at install-time.

I don't think pip allows for that.

Do you have an example of a python module that configures stuff like this at build/install time?

More importantly: maybe ecm was not installed when I installed sagemath. I tried to do something which needed ecm, which resulted in an error "feature not present, please install ecm".

I doubt a couple of calls to `shutil.which()` are more expensive than a single `try: import MODULE`, which we do in oodles. And the first call we need to do it anyway to test the feature, and we can just try first the standard upstream name.

Notes:
 a. we don't run the binary, only see if it's there in path and executable
 b. the feature will not be tested at startup, only the first time it's used

Reply:
> > The executable name never changes so it's wasteful to check it at runtime. When sagelib gets a build system, we'll be able to get rid of the sage-conf variable, too -- ideally the name would be hard-coded at install-time.
> 
> I don't think pip allows for that.
> 
> Do you have an example of a python module that configures stuff like this at build/install time?

Most python modules with C extensions have a bad approximation of `AC_PATH_PROG` and/or `AC_SEARCH_LIBS` in setup.py. For an example of that, look at how hard cvxopt tries to guess your library/include dirs while still getting them wrong.

There was always a better way (autotools, cmake, etc.), but not one that worked well with python. Thankfully, we now have meson-python to fill that gap. NumPy, for example, is using it, and detecting stuff like this at build time. You're CC'ed on @tobiasdiez's PR adding a meson build system for sage. We could handle this with single, standard check at build time. 

> More importantly: maybe ecm was not installed when I installed sagemath. I tried to do something which needed ecm, which resulted in an error "feature not present, please install ecm".

Is that really what you want, or would you rather have ecm pulled in as a dependency when you install the sage component that uses it? These "automagic" dependencies drive our users crazy. Hypothetical scenario:

1. I install sage at home, where I have gmp-ecm installed as a dependency of some other math software
2.  I write a function as an example for a class. It happens to use ECM under the hood without my knowledge because I'm an applied mathematician and what do I know about factoring numbers.
3. Everything works great, i save it, create the slides, upload it to blackboard, whatever
4. I teach the class
5. That night, I get ten emails from students saying my code doesn't work and it says "feature not present, please install ecm"
6. I spend the entire night doing tech support for a bunch of kids' windows laptops

It's OK that features get added by surprise when you install some package, but it's not OK that features get removed by surprise.

> I doubt a couple of calls to `shutil.which()` are more expensive than a single `try: import MODULE`, which we do in oodles. And the first call we need to do it anyway to test the feature, and we can just try first the standard upstream name.

Paraphrasing my mother: N wrongs don't make a right. It's a small improvement, but why not take it?

Reply:
For the record, the meson PR is https://github.com/sagemath/sage/pull/36524. Works completely without sage-conf and sage-setup, and determines the necessary config at build time.

I would say, as a rough guide, binaries should be found at build time and databases/static files should be replaced by python modules. With that sage's features could be simplified (or even partly removed).

Reply:
> > > The executable name never changes so it's wasteful to check it at runtime. When sagelib gets a build system, we'll be able to get rid of the sage-conf variable, too -- ideally the name would be hard-coded at install-time.
> > 
> > 
> > I don't think pip allows for that.
> > Do you have an example of a python module that configures stuff like this at build/install time?
> 
> Most python modules with C extensions have a bad approximation of `AC_PATH_PROG` and/or `AC_SEARCH_LIBS` in setup.py. For an example of that, look at how hard cvxopt tries to guess your library/include dirs while still getting them wrong.

My point stands.

> There was always a better way (autotools, cmake, etc.), but not one that worked well with python. Thankfully, we now have meson-python to fill that gap. NumPy, for example, is using it, and detecting stuff like this at build time. You're CC'ed on @tobiasdiez's PR adding a meson build system for sage. We could handle this with single, standard check at build time.

I unsubscribed to keep my sanity. Moreover, I'm not interested in building the kitchen sink. For sagelib, setuptools builds it just fine, but depending on sage_conf, sage.env, configure-time detection that only happens when you build the kitchen sink, is just taxing me to keep me in your walled-garden. Thanks but, no thanks.

> > More importantly: maybe ecm was not installed when I installed sagemath. I tried to do something which needed ecm, which resulted in an error "feature not present, please install ecm".
> 
> Is that really what you want, or would you rather have ecm pulled in as a dependency when you install the sage component that uses it? These "automagic" dependencies drive our users crazy. Hypothetical scenario:

I'm not talking about ecm in particular. I'm talking about a general principle. I think "ecm" is marked as standard in sage-the-distro and I have it as a dependency on void package. Moreover, not searching for "ecm" at runtime will not make it more likely that your students have it installed, but it will make it more likely to not work if they later install it.

> It's OK that features get added by surprise when you install some package, but it's not OK that features get removed by surprise.


> > I doubt a couple of calls to `shutil.which()` are more expensive than a single `try: import MODULE`, which we do in oodles. And the first call we need to do it anyway to test the feature, and we can just try first the standard upstream name.
> 
> Paraphrasing my mother: N wrongs don't make a right. It's a small improvement, but why not take it?

I don't think it's an improvement, and this PR is already calling `shutil.which()` once, what I say is, if `shutil.which('ecm')` fails, try `shutil.which('gmp-ecm')` before complaining. I'm sure the energy wasted doing that in a year of world-wide use of sagemath will be way less than the energy it would take to have fedora fix their disrespect of upstream.

Reply:
> For the record, the meson PR is #36524. Works completely without sage-conf and sage-setup, and determines the necessary config at build time.
> 
> I would say, as a rough guide, binaries should be found at build time and databases/static files should be replaced by python modules. With that sage's features could be simplified (or even partly removed).

With #37024 I can build sagemath-standard from pypi with no configure step involved and no sage_conf hack.

Reply:
> I unsubscribed to keep my sanity. Moreover, I'm not interested in building the kitchen sink. For sagelib, setuptools builds it just fine, but depending on sage_conf, sage.env, configure-time detection that only happens when you build the kitchen sink, is just taxing me to keep me in your walled-garden. Thanks but, no thanks.

I'm pretty sure we want exactly the same thing. If you ever think I'm concerned about sage-conf, sage-env, or anything under `build/`, then please slow down and read again. The meson build system builds _only_ sagelib. That's the point.

Setuptools cannot build sagelib, unless you happen to run your own distro. Setuptools can't detect C libraries or programs in general; see e.g. cvxopt for what happens when you try. Meson lets us fix that: it gives us a way to build _only_ sagelib, but reliably, and without all of the sage-the-distro scaffolding.

> I don't think it's an improvement, and this PR is already calling `shutil.which()` once, what I say is, if `shutil.which('ecm')` fails, try `shutil.which('gmp-ecm')` before complaining. I'm sure the energy wasted doing that in a year of world-wide use of sagemath will be way less than the energy it would take to have fedora fix their disrespect of upstream.

Again, we don't disagree. I feel the same way about distros renaming executables and breaking consumers. It was my first comment on this ticket.

> With #37024 I can build sagemath-standard from pypi with no configure step involved and no sage_conf hack.

One more case where we want exactly the same thing. I just remade _combinatorial_designs_ as a python package (https://github.com/orlitzky/mols-handbook-data) and am waiting for a repo to make a release (https://github.com/sagemath/sage/issues/37025).

Reply:
> I don't think it's an improvement, and this PR is already calling `shutil.which()` once

hmm, where does this happen?

Reply:
> > I don't think it's an improvement, and this PR is already calling `shutil.which()` once
> 
> hmm, where does this happen?

That's how the `Executable` feature knows it's available. My proposal is to modify this class so one can pass a list of strings instead of just a string. In that case, it would try each option until the first one that works.

Reply:
I never understood the need for sage.features. It seems to papering over a deficiency in Sage package management...